File: scripts/build-backend.py
#!/usr/bin/env python3
"""
Build script for MyACCOBot Python backend using PyInstaller
"""

import os
import sys
import subprocess
import shutil
import platform
from pathlib import Path

# Get the project root directory
PROJECT_ROOT = Path(__file__).parent.parent
SERVER_DIR = PROJECT_ROOT / "Server"
DIST_DIR = PROJECT_ROOT / "dist"
BACKEND_DIST_DIR = SERVER_DIR / "dist"

def get_executable_name():
    """Get platform-specific executable name"""
    system = platform.system().lower()
    machine = platform.machine().lower()
    
    if system == "windows":
        return "myaccobot-backend.exe"
    elif system == "darwin":
        if machine in ["arm64", "aarch64"]:
            return "myaccobot-backend-arm64"
        else:
            return "myaccobot-backend-x64"
    else:  # Linux
        return "myaccobot-backend"

def clean_build():
    """Clean previous builds"""
    print("🧹 Cleaning previous builds...")
    
    dirs_to_clean = [
        BACKEND_DIST_DIR,
        SERVER_DIR / "build",
        SERVER_DIR / "__pycache__",
    ]
    
    for dir_path in dirs_to_clean:
        if dir_path.exists():
            shutil.rmtree(dir_path)
            print(f"   Removed: {dir_path}")

def check_venv():
    """Check if Python virtual environment exists"""
    venv_path = SERVER_DIR / "venv"
    if not venv_path.exists():
        print("❌ Python virtual environment not found!")
        print("   Please run: cd Server && python -m venv venv && source venv/bin/activate && pip install -r requirements.txt")
        sys.exit(1)
    return venv_path

def get_venv_python():
    """Get the Python executable from virtual environment"""
    venv_path = check_venv()
    system = platform.system().lower()
    
    if system == "windows":
        return venv_path / "Scripts" / "python.exe"
    else:
        return venv_path / "bin" / "python"

def install_pyinstaller():
    """Install PyInstaller if not present"""
    print("📦 Checking PyInstaller installation...")
    
    python_exe = get_venv_python()
    
    try:
        # Check if PyInstaller is already installed
        result = subprocess.run(
            [str(python_exe), "-c", "import PyInstaller; print('PyInstaller is installed')"],
            capture_output=True,
            text=True,
            cwd=SERVER_DIR
        )
        
        if result.returncode == 0:
            print("✅ PyInstaller is already installed")
            return
        
        # Install PyInstaller
        print("📦 Installing PyInstaller...")
        subprocess.run(
            [str(python_exe), "-m", "pip", "install", "pyinstaller"],
            check=True,
            cwd=SERVER_DIR
        )
        print("✅ PyInstaller installed successfully")
        
    except subprocess.CalledProcessError as e:
        print(f"❌ Failed to install PyInstaller: {e}")
        sys.exit(1)

def build_executable():
    """Build the Python executable using PyInstaller"""
    print("🔨 Building Python executable...")
    
    python_exe = get_venv_python()
    
    try:
        subprocess.run(
            [str(python_exe), "-m", "PyInstaller", "main.spec"],
            check=True,
            cwd=SERVER_DIR
        )
        print("✅ Python executable built successfully")
        
    except subprocess.CalledProcessError as e:
        print(f"❌ Build failed: {e}")
        sys.exit(1)

def copy_executable():
    """Copy executable to dist directory"""
    print("📋 Copying executable to dist directory...")
    
    executable_name = get_executable_name()
    source_path = BACKEND_DIST_DIR / executable_name
    target_dir = DIST_DIR / "backend"
    target_path = target_dir / executable_name
    
    # Create backend directory in dist
    target_dir.mkdir(parents=True, exist_ok=True)
    
    if source_path.exists():
        shutil.copy2(source_path, target_path)
        print(f"✅ Copied {executable_name} to dist/backend/")
        
        # Make executable on Unix systems
        if platform.system().lower() != "windows":
            os.chmod(target_path, 0o755)
    else:
        print(f"❌ Executable not found at {source_path}")
        sys.exit(1)

def main():
    """Main build function"""
    try:
        print("🚀 Starting backend build process...")
        print(f"   Platform: {platform.system()}")
        print(f"   Architecture: {platform.machine()}")
        print(f"   Executable name: {get_executable_name()}")
        
        clean_build()
        install_pyinstaller()
        build_executable()
        copy_executable()
        
        print("🎉 Backend build completed successfully!")
        print(f"   Executable location: {DIST_DIR / 'backend' / get_executable_name()}")
        
    except Exception as e:
        print(f"💥 Backend build failed: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: scripts/build-backend.js
#!/usr/bin/env node

import { spawn, exec } from 'child_process';
import path from 'path';
import fs from 'fs';
import os from 'os';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const SERVER_DIR = path.join(__dirname, '../Server');
const DIST_DIR = path.join(__dirname, '../dist');
const BACKEND_DIST_DIR = path.join(SERVER_DIR, 'dist');

// Platform-specific executable names
const getExecutableName = () => {
  const platform = os.platform();
  
  if (platform === 'win32') {
    return 'myaccobot-backend.exe';
  } else {
    return 'myaccobot-backend'; // PyInstaller creates the same name for all Unix platforms
  }
};

// Clean previous builds
const cleanBuild = () => {
  console.log('🧹 Cleaning previous builds...');
  
  const dirsToClean = [
    BACKEND_DIST_DIR,
    path.join(SERVER_DIR, 'build'),
    path.join(SERVER_DIR, '__pycache__'),
  ];
  
  dirsToClean.forEach(dir => {
    if (fs.existsSync(dir)) {
      fs.rmSync(dir, { recursive: true, force: true });
      console.log(`   Removed: ${dir}`);
    }
  });
};

// Check if Python virtual environment exists
const checkVenv = () => {
  const venvPath = path.join(SERVER_DIR, 'venv');
  if (!fs.existsSync(venvPath)) {
    console.error('❌ Python virtual environment not found!');
    console.error('   Please run: cd Server && python -m venv venv && source venv/bin/activate && pip install -r requirements.txt');
    process.exit(1);
  }
  return venvPath;
};

// Install PyInstaller if not present
const installPyInstaller = (venvPath) => {
  return new Promise((resolve, reject) => {
    console.log('📦 Checking PyInstaller installation...');
    
    const pipPath = os.platform() === 'win32' 
      ? path.join(venvPath, 'Scripts', 'pip.exe')
      : path.join(venvPath, 'bin', 'pip3');
    
    // Check if PyInstaller is already installed
    exec(`${pipPath} show pyinstaller`, { cwd: SERVER_DIR }, (error, stdout, stderr) => {
      if (!error) {
        console.log('✅ PyInstaller is already installed');
        resolve();
        return;
      }
      
      // Install PyInstaller
      console.log('📦 Installing PyInstaller...');
      exec(`${pipPath} install pyinstaller`, { cwd: SERVER_DIR }, (installError, installStdout, installStderr) => {
        if (installError) {
          console.error('❌ Failed to install PyInstaller:', installError.message);
          reject(installError);
          return;
        }
        console.log('✅ PyInstaller installed successfully');
        resolve();
      });
    });
  });
};

// Build the Python executable
const buildExecutable = (venvPath) => {
  return new Promise((resolve, reject) => {
    console.log('🔨 Building Python executable...');
    
    const pyinstallerPath = os.platform() === 'win32' 
      ? path.join(venvPath, 'Scripts', 'pyinstaller.exe')
      : path.join(venvPath, 'bin', 'pyinstaller');
    
    const child = spawn(pyinstallerPath, ['main.spec'], { 
      cwd: SERVER_DIR,
      stdio: 'inherit'
    });
    
    child.on('close', (code) => {
      if (code === 0) {
        console.log('✅ Python executable built successfully');
        resolve();
      } else {
        console.error(`❌ Build failed with exit code ${code}`);
        reject(new Error(`Build failed with exit code ${code}`));
      }
    });
    
    child.on('error', (error) => {
      console.error('❌ Build process error:', error.message);
      reject(error);
    });
  });
};

// Copy executable to dist directory
const copyExecutable = () => {
  console.log('📋 Copying executable to dist directory...');
  
  const executableName = getExecutableName();
  const sourcePath = path.join(BACKEND_DIST_DIR, executableName);
  const targetPath = path.join(DIST_DIR, 'backend', executableName);
  
  // Create backend directory in dist
  const backendDir = path.join(DIST_DIR, 'backend');
  if (!fs.existsSync(backendDir)) {
    fs.mkdirSync(backendDir, { recursive: true });
  }
  
  if (fs.existsSync(sourcePath)) {
    fs.copyFileSync(sourcePath, targetPath);
    console.log(`✅ Copied ${executableName} to dist/backend/`);
    
    // Make executable on Unix systems
    if (os.platform() !== 'win32') {
      fs.chmodSync(targetPath, '755');
    }
  } else {
    console.error(`❌ Executable not found at ${sourcePath}`);
    throw new Error('Executable not found');
  }
};

// Main build function
const buildBackend = async () => {
  try {
    console.log('🚀 Starting backend build process...');
    console.log(`   Platform: ${os.platform()}`);
    console.log(`   Architecture: ${os.arch()}`);
    console.log(`   Executable name: ${getExecutableName()}`);
    
    cleanBuild();
    const venvPath = checkVenv();
    await installPyInstaller(venvPath);
    await buildExecutable(venvPath);
    copyExecutable();
    
    console.log('🎉 Backend build completed successfully!');
    console.log(`   Executable location: ${path.join(DIST_DIR, 'backend', getExecutableName())}`);
    
  } catch (error) {
    console.error('💥 Backend build failed:', error.message);
    process.exit(1);
  }
};

// Run if called directly
if (import.meta.url === `file://${process.argv[1]}`) {
  buildBackend();
}

export { buildBackend, getExecutableName };
